File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        pos = deque()
        ans = ""

        for x in s:
            if x == "(":
                pos.append(len(ans))
            # when brackets are closed, reverse the string between the brackets
            elif x == ")":
                index = pos.pop()
                tmp = ans[index::]
                ans = ans[0:index] + tmp[::-1]
            else:
                ans += x

        return ans
        

File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        stack<int> pos;
        string ans = "";
        
        for (int i=0; i<s.length(); i++)
        {
            if (s[i] == '(')
            {
                pos.push(ans.length());
            }
            else if (s[i] == ')')
            {
                int index = pos.top();
                pos.pop();
                reverse(ans.begin()+index, ans.end());
            }
            else
            {
                ans += s[i];   
            }
        }
        
        return ans;
    }
};